src/AWS/Lambda/Events/Kafka.hs

```diff
@@ -0,0 +1,249 @@
+import           GHC.Generics           (Generic)
+
+#if MIN_VERSION_aeson(2,0,0)
+import qualified Data.Aeson.Key         as Key
+import qualified Data.Aeson.KeyMap      as KM
+
+keyToText :: Key -> Text
+keyToText = Key.toText
+
+keyFromText :: Text -> Key
+keyFromText = Key.fromText
+#else
+import qualified Data.HashMap.Strict    as KM
+type Key = Text
+
+keyToText :: Key -> Text
+keyToText = id
+
+keyFromText :: Text -> Key
+keyFromText = id
+#endif
+
+-- | Represents an event from either Amazon MSK or a self-managed
+-- Apache Kafka cluster, as the payloads are very similar.
+--
+-- The 'ToJSON' and 'FromJSON' instances on 'Record' perform base64
+-- conversion for you.
+--
+-- See the <https://docs.aws.amazon.com/lambda/latest/dg/with-kafka.html AWS documentation>
+-- for a sample payload.
+data KafkaEvent = KafkaEvent {
+  eventSource      :: !EventSource,
+  -- | Only present when using the "Amazon MSK" event source mapping.
+  eventSourceArn   :: !(Maybe Text),
+  bootstrapServers :: !(NonEmpty Text),
+  -- | The map's keys look like @"${topicName}-${partitionNumber}"@
+  records          :: !(Map Text [Record])
+} deriving (Eq, Show, Generic)
+
+instance FromJSON KafkaEvent where
+  parseJSON = withObject "KafkaEvent" $ \o ->
+    let parseBootstrapServers
+          = maybe (fail "bootstrapServers: empty string") pure
+          . NE.nonEmpty
+          . T.splitOn ","
+    in KafkaEvent
+         <$> o .: "eventSource"
+         <*> o .:! "eventSourceArn"
+         <*> (o .: "bootstrapServers" >>= parseBootstrapServers)
+         <*> o .: "records"
+
+instance ToJSON KafkaEvent where
+  toJSON e = object $
+    [ "eventSource" .= eventSource e
+    , "bootstrapServers" .= T.intercalate "," (NE.toList (bootstrapServers e))
+    , "records" .= records e
+    ] <> maybeToList (("eventSourceArn" .=) <$> eventSourceArn e)
+
+data EventSource
+  = AwsKafka -- ^ @"aws:kafka"@
+  | SelfManagedKafka -- ^ @"SelfManagedKafka"@
+  deriving (Eq, Show, Bounded, Enum, Ord, Generic)
+
+instance FromJSON EventSource where
+  parseJSON = withText "EventSource" $ \case
+    "aws:kafka" -> pure AwsKafka
+    "SelfManagedKafka" -> pure SelfManagedKafka
+    t -> fail $ "unrecognised EventSource: \"" <> show t <> "\""
+
+instance ToJSON EventSource where
+  toJSON = \case
+    AwsKafka -> String "aws:kafka"
+    SelfManagedKafka -> String "SelfManagedKafka"
+
+-- | Convenience alias: most of the time you will parse the records
+-- straight into some app-specific structure.
+type Record = Record' ByteString
+
+-- | Records from a Kafka event. This is 'Traversable', which means
+-- you can do things like parse a JSON-encoded payload:
+--
+-- @
+-- 'traverse' 'decodeStrict' :: 'FromJSON' a => Record ByteString -> Maybe (Record a)
+-- @
+data Record' a = Record {
+  topic :: !Text,
+  partition :: !Int32,
+  offset :: !Int64,
+  timestamp :: !Timestamp,
+  -- | NOTE: there can be multiple headers for a given key.
+  headers :: !(Vector Header),
```

Comment:
Is `Vector` worth the dependency here?  If order is unimportant, it might make sense to do something like `Map Text [ByteString]` which seems like a common approach for HTTP header representation.

Reply:
We indirectly depend on `vector` anyway because `aeson` uses it for its `Array` constructor. I am not particularly fussy about how we represent headers, but if you'd prefer `Map Text (NonEmpty ByteString)` I can rework.

Reply:
I don't want to pull in `MonoidalMap` because its maintenance is slow and it currently depends on `lens` to provide `instance FunctorWithIndex` (and friends). These were factored into a separate `indexed-traversable` package for `lens >=5`.

The use of `Vector` here was suggested by @axman6 because he's a performance hound, the size of the vector is known (because we've parsed as far as `Value` already), and parsing to `[]` loses that length information and introduces a bunch of additional allocations.

However, given that the size of lambda payloads is bounded (by AWS limits) and other event mapping types in this library use `[]`, I will rework for consistency and ergonomics.